analytic_engine/src/row_iter/record_batch_stream.rs

```diff
@@ -77,83 +101,76 @@ pub type SequencedRecordBatchStream = Box<
+/// necessary. Returns the original batch if only a small proportion of the
+/// batch is filtered out.
 fn maybe_filter_record_batch(
     mut sequenced_record_batch: SequencedRecordBatch,
-    filter: &RecordBatchFilter,
-    selected_rows_buf: &mut Vec<bool>,
-) -> Option<SequencedRecordBatch> {
-    if filter.is_empty() {
-        return Some(sequenced_record_batch);
-    }
-
-    // The filter requires the `selected_rows_buf.len() >=
-    // sequenced_record_batch.num_rows()`.
-    selected_rows_buf.resize(sequenced_record_batch.num_rows(), true);
-    let num_selected_rows = filter.filter(
-        &sequenced_record_batch.record_batch,
-        selected_rows_buf.as_mut_slice(),
-    );
-
-    trace!(
-        "filter record batch, selected_rows:{}, origin_rows:{}",
-        num_selected_rows,
-        sequenced_record_batch.num_rows()
-    );
-
-    // No row is selected.
-    if num_selected_rows == 0 {
-        return None;
-    }
-
-    if num_selected_rows
-        > sequenced_record_batch.num_rows() / REBUILD_FILTERED_RECORD_BATCH_MAGNIFICATION
-    {
-        // just use the original record batch because only a small proportion is
-        // filtered out.
-        return Some(sequenced_record_batch);
-    }
+    predicate: Arc<dyn PhysicalExpr>,
+) -> Result<Option<SequencedRecordBatch>> {
+    let record_batch = sequenced_record_batch.record_batch.as_arrow_record_batch();
+    let filter_array = predicate
+        .evaluate(record_batch)
+        .map(|v| v.into_array(record_batch.num_rows()))
+        .context(FilterExec)?;
+    let selected_rows = filter_array
+        .as_any()
+        .downcast_ref::<BooleanArray>()
+        .context(DowncastBooleanArray {
+            data_type: filter_array.as_ref().data_type().clone(),
+        })?;
 
-    // select the rows according to the filter result.
-    if let Err(e) = sequenced_record_batch
+    sequenced_record_batch
         .record_batch
-        .select_data(selected_rows_buf.as_slice())
-    {
-        error!(
-            "Fail to select record batch, data:{:?}, selected_rows:{:?}, err:{}",
-            sequenced_record_batch, selected_rows_buf, e,
-        );
-    }
+        .select_data(selected_rows)
+        .context(SelectBatchData)?;
 
-    Some(sequenced_record_batch)
+    sequenced_record_batch
+        .record_batch
+        .is_empty()
+        .not()
+        .then_some(Ok(sequenced_record_batch))
+        .transpose()
 }
 
 /// Filter the sequenced record batch stream by applying the `predicate`.
 /// However, the output record batches is not ensured to meet the requirements
 /// of the `predicate`.
 pub fn filter_stream(
     origin_stream: SequencedRecordBatchStream,
+    input_schema: ArrowSchemaRef,
     predicate: &Predicate,
-) -> SequencedRecordBatchStream {
-    if predicate.exprs().is_empty() {
-        return origin_stream;
+) -> Result<SequencedRecordBatchStream> {
+    let filter = combine_filters(predicate.exprs());
+    if let Some(filter) = filter {
```

Comment:
Yes, it can remove long if condition.

But it is a little awkward to introduce `unwrap`, sometimes we can use `unwrap_or_default` in the similar conditions, unfortunately datafuison's `Expr` do not implement `Default`.

I am not sure it is worth to refactor.

Reply:
Addressed.